crates/noirc_evaluator/src/ssa/node.rs

```diff
@@ -0,0 +1,948 @@
+    //  pub max_bits: u32,                  //max possible bit size of the expression
+    //  pub max_value: Option<BigUInt>,     //maximum possible value of the expression, if less than max_bits
+    pub witness: Option<Witness>,
+    pub parent_block: arena::Index,
+}
+
+impl Variable {
+    pub fn get_root(&self) -> arena::Index {
+        match self.root {
+            Some(r) => r,
+            _ => self.id,
+        }
+    }
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Copy, Clone, Debug)]
+pub enum ObjectType {
+    native_field,
+    // custom_field(BigUint), //TODO requires copy trait for BigUint
+    boolean,
+    unsigned(u32), //bit size
+    signed(u32),   //bit size
+    //custom(u32),   //user-defined struct, u32 refers to the id of the type in...?todo
+    //array(ObjectType),  TODO we should have primitive type and composite types
+    //TODO big_int
+    //TODO floats
+    none, //not an object
+}
+
+impl ObjectType {
+    fn is_signed(&self) -> bool {
+        matches!(self, ObjectType::signed(_))
+    }
+
+    fn is_unsigned(&self) -> bool {
+        matches!(self, ObjectType::unsigned(_))
+    }
+
+    fn is_field(&self) -> bool {
+        matches!(
+            self,
+            ObjectType::native_field //| ObjectType::custom_field
+        )
+    }
+
+    // pub fn from_type(t: noirc_frontend::Type) -> ObjectType {
+    //     match t {
+    //         noirc_frontend::Type::FieldElement(_) => ObjectType::native_field,
+    //         noirc_frontend::Type::Array(_,_,_) => ObjectType::none, //TODO
+    //         noirc_frontend::Type::Integer(_ftype, sign, bit_size) => {
+    //             match sign {
+    //                 //todo FieldElementType?
+    //                 noirc_frontend::Signedness::Signed => ObjectType::signed(bit_size),
+    //                 noirc_frontend::Signedness::Unsigned => ObjectType::unsigned(bit_size),
+    //             }
+    //         }
+    //         noirc_frontend::Type::Bool => ObjectType::boolean,
+    //         _ => ObjectType::none, //todo Error,Unspecified, Unknown,Unit
+    //     }
+    // }
+
+    pub fn get_type_from_object(obj: &Object) -> ObjectType {
+        match obj {
+            Object::Arithmetic(_) => {
+                todo!();
+                //ObjectType::native_field
+            }
+            Object::Array(_) => {
+                todo!();
+                //ObjectType::none
+            }
+            Object::Constants(_) => ObjectType::native_field, //TODO
+            Object::Integer(i) => ObjectType::unsigned(i.num_bits), //TODO signed or unsigned?
+            Object::Linear(_) => {
+                todo!();
+                //ObjectType::native_field
+            }
+            Object::Null => ObjectType::none,
+        }
+    }
+
+    pub fn bits(&self) -> u32 {
+        match self {
+            ObjectType::boolean => 1,
+            ObjectType::native_field => FieldElement::max_num_bits(), //TODO is it the correct value?
```

Comment:
I believe so, this is the maximum number of bits needed to represent any field element. So 254 for bn254

Reply:
OK